pkg/app/pipedv1/plugin/sourcecloner/cloner.go

```diff
@@ -0,0 +1,82 @@
+// Copyright 2024 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sourcecloner
+
+import (
+	"context"
+	"os"
+	"os/exec"
+	"sync"
+
+	"github.com/pipe-cd/pipecd/pkg/app/piped/deploysource"
+	"github.com/pipe-cd/pipecd/pkg/git"
+)
+
+var _ deploysource.SourceCloner = (*Cloner)(nil)
+
+var gitPath = sync.OnceValues(func() (string, error) {
+	path, err := exec.LookPath("git")
+	if err != nil {
+		return "", err
+	}
+	return path, nil
+})
+
+// Cloner is a source cloner for git repositories.
+type Cloner struct {
+	gitPath      string
+	remotePath   string
+	revision     string
+	revisionName string
+}
```

Comment:
@Warashi I have an idea with this sourcecloner package 🤔 
Basically, our plan is to use the `piped/deploysource` as the portal for all git resource operations (indeed, the name is deploysource and it emits the internal using git stub).
I think this newly created package `sourcecloner` should follow that approach and more. What package dose is to fetch the already cloned source (prepared by deploysource), then I think we should emits the internal using git as same as we do in deploysource.

Reply:
Instead of a `Clone` interface, which could lead to confusion (if the reviewer doesn't know we don't clone or any, just fetch the cloned local source prepared by deploysource), how about to reuse the SourceCloner interface. aka this: https://github.com/pipe-cd/pipecd/blob/master/pkg/app/piped/deploysource/sourcecloner.go#L24-L28
I can see we have some implementation of this interface looks pretty fits our use case here: 
https://github.com/pipe-cd/pipecd/blob/master/pkg/app/piped/deploysource/sourcecloner.go#L69-L94

Reply:
Oops, I missed the localSourceCloner. Thank you!
I think we can use this, but I wonder why this `Clone` implementation only invokes the Copy method of the internal git repo.
If the internal git repo's current HEAD doesn't match localSourceCloner's revision, this Clone method will give the wrong results.
Or have I misunderstood the `Revision()` of SourceCloner?

Reply:
@Warashi That's a good catch 👍 I think up to now, the localSourceCloner not yet raised any issues because its caller doesn't require any other revision than the HEAD 🤔 So, basically I think it's a bug potential logic and should be fixed, but as far as I investigated, the only place which we use this localSourceCloner is `planpreview/builder` (ref: this code: https://github.com/pipe-cd/pipecd/blob/master/pkg/app/piped/planpreview/builder.go#L228), which basically only request for the HEAD of the current branch - due to the feature plan preview logic.
So, my point is to fix this localSourceCloner and use it for whatever "NOT git action pull directly but pull/fetch from a local source". WDYT? 🤔

Reply:
My point is: we should not create too many interfaces/packages that do similar tasks (clone source from git), or it will lead to confusion for the caller side. In this case, it would be better to have only one SourceCloner interface, one pull/fetch to the outer git, one pull/fetch to the local repo that was pulled by the previous implementation.

Reply:
I agree with you.
I'll fix the localSourceCloner and use it.